pkg/operator/staticpod/controller/installerstate/installer_state_controller.go

```diff
@@ -67,13 +76,36 @@ func (c *InstallerStateController) sync(ctx context.Context, syncCtx factory.Syn
 		return err
 	}
 
+	masterRevisions := make(map[string][]*v1.Pod)
+	installerHighestRunningRevision := make(map[string]int)
+	for _, pod := range pods.Items {
+		p := pod
+		masterRevisions[pod.Spec.NodeName] = append(masterRevisions[pod.Spec.NodeName], &p)
+	}
+	// find the highest revision of a non-pending pod on each master node
+	for masterNode, pods := range masterRevisions {
+		maxRunningRev := 0
+		for _, pod := range pods {
+			if pod.Status.Phase != v1.PodPending || pod.Status.StartTime == nil {
+				rev, err := installerNameToRevision(pod.Name)
+				if err != nil {
+					return err
+				}
+				if rev > maxRunningRev {
+					maxRunningRev = rev
+				}
+			}
+		}
+		installerHighestRunningRevision[masterNode] = maxRunningRev
+	}
+
 	// collect all startingObjects that are in pending state for longer than maxToleratedPodPendingDuration
 	pendingPods := []*v1.Pod{}
 	for _, pod := range pods.Items {
 		if pod.Status.Phase != v1.PodPending || pod.Status.StartTime == nil {
 			continue
 		}
-		if c.timeNowFn().Sub(pod.Status.StartTime.Time) >= maxToleratedPodPendingDuration {
+		if rev, _ := installerNameToRevision(pod.Name); c.timeNowFn().Sub(pod.Status.StartTime.Time) >= maxToleratedPodPendingDuration && rev >= installerHighestRunningRevision[pod.Spec.NodeName] {
```

Comment:
The way you are building your list of pending pods is that you are taking all the pending pods that are more recent than the latest completed installer pod, but I don't think we need to report the status of all the pending pods, only the latest one should be enough since users will most likely only be interested about what happened to the latest revision.

For that purpose, it might be better to just build a map of the latest installer pod (no matter its state) per node and then pass it on to this loop that builds a list of pods that are in a pending state for longer than maxToleratedPodPendingDuration.

Reply:
> only the latest one for each node should be enough since users will most likely only be interested about what happened to the latest revision.

Good point. The latest installer pod invalidates all pending installer pods of a lower revision.

> For that purpose, it might be better to just build a map of the latest installer pod (no matter its state) per node and then pass it on to this loop that builds a list of pods that are in a pending state for longer than maxToleratedPodPendingDuration.

A completed pod is a base for deciding whether there are resources of a given revision successfully written to a master node. Existence of two or more pending pods with a higher revision than the latest completed revision might signify there's someone wrong happening on a master node. Question is whether this case is important for reporting. When the installer_state_controller reports Degraded state, there will always be at least one pending installer pod. Each pending pod is reported through a [separate condition](https://github.com/knelasevero/library-go/blob/702c72f4495d8e74dec40601b4520d27f53b7e6f/pkg/operator/staticpod/controller/installerstate/installer_state_controller.go#L197). Will reporting two or more pending pods for the same master node help in debugging? Will it provide some additional value to an admin? If not, let's simplify the logic and report only pending pods of the latest revision. I wonder whether the current/original implementation took this reasoning into consideration. When a must-gather is collected all the pending pods are collected as well. Though, all the conditions are logged by an operator. So having the additional pods reported in the conditions might give us more detailed historical background of what was happening to the installer pods after they eventually either completed or were deleted but no longer collected by must-gather.

Reply:
> Though, all the conditions are logged by an operator. So having the additional pods reported in the conditions might give us a more detailed historical background of what was happening to the installer pods after they eventually either completed or were deleted but no longer collected by must-gather.

I can buy that argument, having historical background makes sense. It might not end up being useful for debugging, but at least it could for example help us track down when slowness first started to appear.